### PR TITLE
support new API

### PR DESCRIPTION
**Pull request by jleni**
- getAddress
- appInfo (works for any app including BOLOS dashboard)
- improved error messages

Related to: 
https://github.com/ZondaX/ledger-cosmos-js/issues/15
https://github.com/cosmos/voyager/issues/2299
https://github.com/cosmos/voyager/issues/2298
https://github.com/cosmos/voyager/issues/1962

### CODE REVIEW

File: src/ledger-app.js [hunk 1]
```diff
@@ -33,6 +33,7 @@ const INS_GET_VERSION = 0x00;
 const INS_PUBLIC_KEY_SECP256K1 = 0x01;
```
Comment (fedekunze) [inline src/ledger-app.js]: do we need this anymore ?

File: src/ledger-app.js [hunk 2]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+
+            result["return_code"] = error_code_data[0] * 256 + error_code_data[1];
+            result["error_message"] = errorMessage(result["return_code"]);
+
+            if (result.return_code !== 0x9000) {
+                return
+            }
+
+            if (apduResponse[0] !== 1) {
+                result["error_message"] = "Invalid response - format ID not recognized";
```
Comment (fedekunze) [inline src/ledger-app.js]: >`format ID not recognized`

what does that mean ?

Comment (fedekunze) [reply src/ledger-app.js]:
```suggestion
                result["error_message"] = "format ID not recognized";
```

File: src/ledger-app.js [hunk 3]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+                return result
+            }
+
+            var idx = 1;
+            const appNameLen = apduResponse[idx];
+            idx++;
+            result["appName"] = apduResponse.slice(idx, idx+appNameLen).toString('ascii');
+            idx += appNameLen;
+
+            const appVerLen=apduResponse[idx];
```
Comment (fedekunze) [inline src/ledger-app.js]:
```suggestion
            const appVersionLen=apduResponse[idx];
```

File: src/ledger-app.js [hunk 4]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+                return
+            }
+
+            if (apduResponse[0] !== 1) {
+                result["error_message"] = "Invalid response - format ID not recognized";
+                return result
+            }
+
+            var idx = 1;
+            const appNameLen = apduResponse[idx];
```
Comment (fedekunze) [inline src/ledger-app.js]: this should always be `1`:
```suggestion
            const appNameLen = apduResponse[1];
```

File: src/ledger-app.js [hunk 5]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+            if (result.return_code !== 0x9000) {
+                return
+            }
+
+            if (apduResponse[0] !== 1) {
+                result["error_message"] = "Invalid response - format ID not recognized";
+                return result
+            }
+
+            var idx = 1;
```
Comment (fedekunze) [inline src/ledger-app.js]:
```suggestion
```

File: src/ledger-app.js [hunk 6]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+            const appNameLen = apduResponse[idx];
+            idx++;
```
Comment (fedekunze) [inline src/ledger-app.js]:
```suggestion
            let idx = 2 + appNameLen;
```

File: src/ledger-app.js [hunk 7]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+            result["appName"] = apduResponse.slice(idx, idx+appNameLen).toString('ascii');
```
Comment (fedekunze) [inline src/ledger-app.js]:
```suggestion
            result["appName"] = apduResponse.slice(2, idx).toString('ascii');
```

File: src/ledger-app.js [hunk 8]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+            idx += appNameLen;
```
Comment (fedekunze) [inline src/ledger-app.js]:
```suggestion
```

File: src/ledger-app.js [hunk 9]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+
+            const appVerLen=apduResponse[idx];
+            idx++;
+            result["appVersion"] = apduResponse.slice(idx, idx+appVerLen).toString('ascii');
```
Comment (fedekunze) [inline src/ledger-app.js]:
```suggestion
            result["appVersion"] = apduResponse.slice(idx, idx+appVersionLen).toString('ascii');
```

File: src/ledger-app.js [hunk 10]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
...
+            idx+=appVerLen;
```
Comment (fedekunze) [inline src/ledger-app.js]:
```suggestion
            idx+=appVersionLen;
```

File: src/ledger-app.js [hunk 11]
```diff
@@ -307,4 +309,76 @@ LedgerApp.prototype.showAddress = function (hrp, path) {
         process_error_response);
 };
 
+LedgerApp.prototype.getAddress = function (hrp, path) {
```
Comment (fedekunze) [inline src/ledger-app.js]:
```suggestion
LedgerApp.prototype.getAddressAndPubkey = function (hrp, path) {
```

File: tests/tests-node.js [hunk 12]
```diff
@@ -398,6 +403,55 @@ describe('showAddress', function () {
             });
     });
     it('return_code is 0x9000', function () {
+        console.log("Error code 0x%s: %s ", response.return_code.toString(16), response.error_message);
+        expect(response.return_code).to.equal(0x9000);
+    });
+});
+
+describe('getAddress', function () {
```
Comment (fedekunze) [inline tests/tests-node.js]:
```suggestion
describe('getAddressAndPubkey', function () {
```

File: tests/tests-node.js [hunk 13]
```diff
@@ -398,6 +403,55 @@ describe('showAddress', function () {
...
+    let response;
+    // call API
+    before(function () {
+        this.timeout(LONG_TIMEOUT);
+        return comm.create_async(LONG_TIMEOUT, true).then(
+            function (comm) {
+                let app = new ledger.App(comm);
+
+                let path = [44, 118, 5, 0, 3];
+                return app.getAddress("cosmos", path).then(function (result) {
```
Comment (fedekunze) [inline tests/tests-node.js]:
```suggestion
                return app.getAddressAndPubkey("cosmos", path).then(function (result) {
```

File: tests/tests-node.js [hunk 14]
```diff
@@ -29,7 +29,7 @@ browser = false;
 const TIMEOUT = 1000;
 const LONG_TIMEOUT = 45000;
 const EXPECTED_MAJOR = 1;
-const EXPECTED_MINOR = 1;
+const EXPECTED_MINOR = 3;
 const EXPECTED_PATCH = 0;
 
 describe('get_version', function () {
```
Comment (fedekunze) [inline tests/tests-node.js]: maybe we should use camelCase here for consistency ?

File: package.json [hunk 15]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ledger-cosmos-js",
-  "version": "0.0.1",
+  "version": "1.0.0",
   "description": "Node API for Ledger Nano S - Tendermint/Cosmos App",
```
Comment (fedekunze) [inline package.json]:
```suggestion
  "description": "Node API for Ledger Nano S Cosmos App",
```

File: src/ledger-app.js (hunk 1 above)
Comment (jleni) [reply src/ledger-app.js]: I just kept it for backward compatibility. Once voyager upgrades, we can remove it

Comment (jleni) [reply src/ledger-app.js]: we should also inform any third parties so they are aware of the change

File: tests/tests-node.js (hunk 14 above)
Comment (jleni) [reply tests/tests-node.js]: I tried to keep this similar to the firmware/app so it is easier to track the functionality in C later.
We can refactor if it is really necessary.

Comment (jleni) [reply tests/tests-node.js]: Sorry, that comment was with respect to the INS constants.
I renamed these ones.

File: src/ledger-app.js (hunk 1 above)
Comment (fedekunze) [reply src/ledger-app.js]: ok let's create an issue then

**Pull request merged**